### DISCUSSION TITLE
NextJS (Auth.js) example with PostgresAdapter

### DISCUSSION THREAD

**Discussion by volodalexey** [vitaly-t/pg-promise#895] [General]
In NextJS (Auth.js) we have [following example](https://authjs.dev/reference/adapter/pg):
```typescript
import NextAuth from "next-auth"
import { PostgresAdapter } from "@auth/pg-adapter"
import { Pool } from "pg"

const pool = new Pool({
  host: "localhost",
  user: "database-user",
  max: 20,
  idleTimeoutMillis: 30000,
  connectionTimeoutMillis: 2000
})

export default NextAuth({
  adapter: PostgresAdapter(pool),
  providers: []
})
```
I want to use `pg-promise` instead.
I found [singletone approach here](https://stackoverflow.com/a/34427278/5431545), however when I try to use internal `$pool` of `pg-promise` it is not the same as `Pool` from pg:
```typescript
import { PostgresAdapter } from "@auth/pg-adapter"
import { getDB } from "../../../../db/db";

export const options: NextAuthOptions = {
  adapter: PostgresAdapter(getDB().db.$pool),
}
```
I see following TS error:
```
Argument of type 'IPool' is not assignable to parameter of type 'Pool'.
  Property 'query' is missing in type 'IPool' but required in type 'Pool'.
```

Under the hood `pg-adapter` calls `query` [from pg pool](https://github.com/nextauthjs/next-auth/blob/main/packages/adapter-pg/src/index.ts#L152).
Any ideas how to use `pg-promise` next-auth within adapter?

**Comment by vitaly-t** [comment]
`IPool` skips `query` declaration on purpose, so nobody tries to execute queries directly on the pool, as it contradicts the general approach for handling queries in this library. However, this does create an issue when it comes to integration with the base Pool, like this, since that library requires direct access to the pool.

So what this all means, in your case, just recast the type, because apart from the declaration, the code is correct:

```ts
export const options: NextAuthOptions = {
  adapter: PostgresAdapter(getDB().db.$pool as any as Pool),
}
```
